Implement sound notification for Firefox

GhoulofGSG9 opened the pull request:
Implement notification sounds for Firefox

Open To-do:

- [x] ~~Implement custom sound support~~

GhoulofGSG9 commented:
Oh great, the issue with the file selector is related to a 10-year-old Firefox bug https://bugzilla.mozilla.org/show_bug.cgi?id=1292701 💀 

So would need to move that option into the extension settings. @cdelaforge Do you have any input on this?

cdelaforge commented:
That's not why I hadn't implemented it on Firefox, I just couldn't find a way to do the same thing as on Chrome.
Does everything else seem possible?
It's annoying, I used to have duplicate options in this popup and in the extension settings, and a few weeks ago I got fed up with all this duplication of code and deleted everything from the extension settings :(
Can you think of a simple way of managing this without duplicating everything in both places?

GhoulofGSG9 commented:
@cdelaforge Let's add general sound support first.

Then I'll look into opening the OptionsView element into an option page tab. As that will require a bit of refactoring and a way to sync state between the option page and popup page (using e.g. a broadcastchannel), I will postpone working on that until the weekend.

GhoulofGSG9 commented:
Okay, updated PR to include changes to fully support custom sounds at Firefox.

Overview of changes:
- Implemented full support for sound notifications at Firefox
  - Opinionated UX change: Toggling the custom sounds slider directly opens the file selection, so users don't have to click the Upload MP3 button. They may still use it to change their custom sound
  - At Firefox, changing the custom sounds option is limited to the context of the extension option context as I could not figure out any non-confusing UX flow to switch between popup and options menu
- Replaces About Tab in extensions settings with a General Tab displaying the OptionsViews
- Ensure states between OptionsViews stay synced between popup and options context
  - Added new hook useSyncedState that ensures states are synced across windows via broadcastchannels
- Refactor OptionsView and all child components to be fully implemented using typescript
  - Added tsconfig for proper linting and type checking
    - This config uses the same options as the parcel transformer and some best practices. Some options may be open to debate
  - Updated sailswind config to include typescript files
- Refactored browser to use Mozilla's web extensions polyfill lib. Parcel did polyfill `chrome` but this way we have proper type support. **Ensure to run `yarn install`!**

GhoulofGSG9 commented:
Rebased on updated main

GhoulofGSG9 commented:
Have been testing this and my other changes over the weekend, and it should be ready to be merged.

GhoulofGSG9 commented:
Rebased on main branch to resolve conflicts ;)
